Japanese Kana language support

thinkyhead opened the pull request:
Derived from http://www.thingiverse.com/thing:664397

The author has licensed the font and translations CC-BY-NC. If this is incompatible with GPL (due to the commercial-use restriction), we can ask the author to revise it, or possibly assert GPL rights (modifications must be contributable to the original source). I leave this question to the gatekeepers.

thinkyhead committed:
Japanese Kana language support

- Derived from http://www.thingiverse.com/thing:664397

thinkyhead committed:
Fix typo in MSG_POSITION_UNKNOWN

thinkyhead committed:
Load 6x9 with Kana also

thinkyhead committed:
Merge remote-tracking branch 'upstream/Development' into japanese_kana_language

AnHardt commented:
@thinkyhead 
I hope this will not crash to many of my branches.

Do you have complete kana and ore kanji font? I have a branch where i made a marlin-special-character font mapped from 1 to 10 or so. This makes possible to use standard fonts an bring back the so very missed 'ö' and 'ü'.
On the other side the Unicode solution in ru.h is very appealing to me but needs more experiments to generalice this for more languages.

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead commented:
@AnHardt
I believe it should dovetail okay, just take care in your merges.

The font included here is purely Kana for graphical display (and finally, Kana for HD44780 - making sure they have the JP display.) Some Kanji may be in there as a conventional aspect of Kana, but I don't know. I've made the Kana font a separate file which made it best to just separate all the fonts for smaller includes (should be a pretty quick post-merge cleanup).

This came along as a fluke, so we can hope that more Japanese (and Chinese, Korean...) users will come along with translations soon too. It will all be terribly unreadable, with all those hex characters. I wonder if we can actually use some kind of literals... I suppose he would have if he could have.

thinkyhead committed:
Make sure user specifies Japanese HD44780

AnHardt commented:
This 
https://github.com/thinkyhead/Marlin/commit/e24d9d11c2adca51bcac04c22fb24d287012ff61#diff-b52dae9bb2142eed11f520c911352dcdR17
shold be: #include "language_jp_2.h" // DOGM
And this
https://github.com/thinkyhead/Marlin/commit/e24d9d11c2adca51bcac04c22fb24d287012ff61#diff-7982dd1a72ec216967121897d03e544bR30
could be '// jp    Japanese' or // 'jp    Japanese Kana / Kanji'

AnHardt commented:
@thinkyhead 
Ah! Sorry. Now i understand. You reduced the original idea to Kana for now.

How about not making half a dozen new fontfiles.h but give the fonts theyr own names and select them abuot here? https://github.com/thinkyhead/Marlin/commit/e24d9d11c2adca51bcac04c22fb24d287012ff61#diff-41c2dba4a9d48a52fb2bbb77e8f1c583R64

AnHardt commented:
@thinkyhead 
This will compile better when the #includes in language_jp.h match the actual filenames.

AnHardt committed:
Merge pull request #2 from thinkyhead/japanese_kana_language

Japanese kana language

AnHardt committed:
Match #includes with existing filenames in language_jp.h

thinkyhead committed:
Merge pull request #3 from AnHardt/thinky_kana

Match #includes with existing filenames in language_jp.h

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead committed:
Fix: incorrect file include for display

thinkyhead committed:
Move JP comments to headers

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead commented on Marlin/dogm_font_data_6x10_marlin_jp.h:
```diff
@@ -0,0 +1,189 @@
+/*
+  Fontname : Japanese Original font, Version 0.1
+  Author   : http://www.thingiverse.com/ichibey
+  License  : Creative Commons BY-NC
```
Appears this is incompatible with GPLv3. Asking permission.

AnHardt commented:
@thinkyhead 
This font is unbelievable tricky. He only uses a very limited and selected number of kanji symbols. I see absolute no chance to make the input for this in utf8. General fonts for Kanji must be unbelievable big and are not very well suitable for embedded systems like our small arduinos.
Oh, and tell him he could use another 8 symbols.

At the moment i'm working on:
a. Isolate the marlin symbols in its on font. Mapped them to the low numbers. Made a output function what switches fonts when symbol-number is lower then ' ' (done and working fine)
b. for that i made the fontswich-function remember the last font. (done and working)
c. exchanged strlen() with your utf8 version and made the str-output functions count the space used. (works)
d. made mapper functions and tables for utf8 to several fonts. The concept is from the, then removed, Russian LCD-files. and will replace the STR_ae ... defines in language.h (done, still faulty)
e. build the fonts of the different types of Hitachis for u8g.lib. (done)
f. still missing a Cyrillic ISO10646 font in the right size. But the font of the hitachis works well.
g. putting all together at the right place. (working on it)
h. test 6*5 combinations of fonts and mappers with at least 4 languages with different input concepts. (this will last for a while)
i. Make documentation. (to be done)
j find a way to merge this widely spreaded changes in dogm_lcd_implementation.h, ultralcd_implementation_hitachi_HD44780.h, ultralcd.cpp, the fonts, Configuration.h, language.h,
and the language files.

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead commented:
Do you mean, there are another 8 symbols available which he may use to improve his selection?

Sounds like you are getting closer to something comprehensive. I look forward to the results!

AnHardt commented:
@thinkyhead 
Yes. Until now he has the marlin symbols in his font. They will not be needed anymore.

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream

AnHardt commented:
My new version is described in #1492.

AnHardt commented:
@thinkyhead 

> Appears this is incompatible with GPLv3. Asking permission.

Have you got an answer?

thinkyhead commented:
@AnHardt Well… Here's what I have. I need help in translation and communication…

英語が苦手なので翻訳サイトを利用しています。

日本語化したmarlinについてです。
日本人としては、恥ずかしいのですが、現在の日本製の3Dプリンタに於いて、marlinが使用されていながらそれが明示されていない物やソースコードを公開していない物が存在しています。

その為、商用利用不可の記載として、コメント文に商用利用する場合は連絡する様に明示してThingiverseへデータをアップロードしています。

商用利用の問い合せが有った場合には、marlinが使用されている事の明示とソースコードの開示を行えば許可しますので、実質的には商用利用可能とするつもりです。

AnHardt commented:
Tried goggle-translator, Babylon and some other translators. But they only paint a very vague picture what he could have said.
Any other Japanese out there?

mmccool commented:
Let me give it a stab (I live in Japan; my Japanese is not super-fluent, though).

> $B1Q8l$,6l<j$J$N$GK]Lu%5%$%H$rMxMQ$7$F$$$^$9!#(BBecause my English is bad I am using a translated web site.  
> 
> $BF|K\8l2=$7$?(Bmarlin$B$K$D$$$F$G$9!#(BRegarding the Japanese language translation of Marlin...
> 
> More to follow, next sentence is a tough run-on one, need to dig out the dictionary, too.  It's past midnight here though so I'll tackle it in the morning.  Something about wanting to use Marlin in Japanese-made 3D printers....

Michael McCool

> On 2015/03/14, at 21:27, Scott Lahteine notifications@github.com wrote:
> 
> @AnHardt Well$B!D(B Here's what I have. I need help in translation and communication$B!D(B
> 
> $B1Q8l$,6l<j$J$N$GK]Lu%5%$%H$rMxMQ$7$F$$$^$9!#(B
> 
> $BF|K\8l2=$7$?(Bmarlin$B$K$D$$$F$G$9!#(B
> $BF|K\?M$H$7$F$O!"CQ$:$+$7$$$N$G$9$,!"8=:_$NF|K\@=$N(B3D$B%W%j%s%?$K1w$$$F!"(Bmarlin$B$,;HMQ$5$l$F$$$J$,$i$=$l$,L@<($5$l$F$$$J$$J_$d%=!<%9%3!<%I$r8x3+$7$F$$$J$$J_$,B8:_$7$F$$$^$9!#(B
> 
> $B$=$N0Y!">&MQMxMQIT2D$N5-:\$H$7$F!"%3%a%s%HJ8$K>&MQMxMQ$9$k>l9g$OO"Mm$9$kMM$KL@<($7$F(BThingiverse$B$X%G!<%?$r%"%C%W%m!<%I$7$F$$$^$9!#(B
> 
> $B>&MQMxMQ$NLd$$9g$;$,M-$C$?>l9g$K$O!"(Bmarlin$B$,;HMQ$5$l$F$$$k;v$NL@<($H%=!<%9%3!<%I$N3+<($r9T$($P5v2D$7$^$9$N$G!"<B<AE*$K$O>&MQMxMQ2DG=$H$9$k$D$b$j$G$9!#(B
> 
> $B!=(B
> Reply to this email directly or view it on GitHub.

mmccool commented:
$BF|K\?M$H$7$F$O!"CQ$:$+$7$$$N$G$9$,!"8=:_$NF|K\@=$N(B3D$B%W%j%s%?$K1w$$$F!"(Bmarlin$B$,;HMQ$5$l$F$$$J$,$i$=$l$,L@<($5$l$F$$$J$$J_$d%=!<%9%3!<%I$r8x3+$7$F$$$J$$J_$,B8:_$7$F$$$^$9!#(B

As for Japanese people, it's embarassing but, in modern Japanese-made printers, while Marlin is being used that fact is not made explicit, and the existence of the source code is not made public.

> On 2015/03/15, at 0:30, Michael McCool michael.d.mccool@gmail.com wrote:
> 
> Let me give it a stab (I live in Japan; my Japanese is not super-fluent, though).
> 
> > $B1Q8l$,6l<j$J$N$GK]Lu%5%$%H$rMxMQ$7$F$$$^$9!#(BBecause my English is bad I am using a translated web site.  
> > 
> > $BF|K\8l2=$7$?(Bmarlin$B$K$D$$$F$G$9!#(BRegarding the Japanese language translation of Marlin...
> > 
> > More to follow, next sentence is a tough run-on one, need to dig out the dictionary, too.  It's past midnight here though so I'll tackle it in the morning.  Something about wanting to use Marlin in Japanese-made 3D printers....
> 
> Michael McCool
> 
> > On 2015/03/14, at 21:27, Scott Lahteine notifications@github.com wrote:
> > 
> > @AnHardt Well$B!D(B Here's what I have. I need help in translation and communication$B!D(B
> > 
> > $B1Q8l$,6l<j$J$N$GK]Lu%5%$%H$rMxMQ$7$F$$$^$9!#(B
> > 
> > $BF|K\8l2=$7$?(Bmarlin$B$K$D$$$F$G$9!#(B
> > $BF|K\?M$H$7$F$O!"CQ$:$+$7$$$N$G$9$,!"8=:_$NF|K\@=$N(B3D$B%W%j%s%?$K1w$$$F!"(Bmarlin$B$,;HMQ$5$l$F$$$J$,$i$=$l$,L@<($5$l$F$$$J$$J_$d%=!<%9%3!<%I$r8x3+$7$F$$$J$$J_$,B8:_$7$F$$$^$9!#(B
> > 
> > $B$=$N0Y!">&MQMxMQIT2D$N5-:\$H$7$F!"%3%a%s%HJ8$K>&MQMxMQ$9$k>l9g$OO"Mm$9$kMM$KL@<($7$F(BThingiverse$B$X%G!<%?$r%"%C%W%m!<%I$7$F$$$^$9!#(B
> > 
> > $B>&MQMxMQ$NLd$$9g$;$,M-$C$?>l9g$K$O!"(Bmarlin$B$,;HMQ$5$l$F$$$k;v$NL@<($H%=!<%9%3!<%I$N3+<($r9T$($P5v2D$7$^$9$N$G!"<B<AE*$K$O>&MQMxMQ2DG=$H$9$k$D$b$j$G$9!#(B
> > 
> > $B!=(B
> > Reply to this email directly or view it on GitHub.

mmccool commented:
> $B$=$N0Y!">&MQMxMQIT2D$N5-:\$H$7$F!"%3%a%s%HJ8$K>&MQMxMQ$9$k>l9g$OO"Mm$9$kMM$KL@<($7$F(BThingiverse$B$X%G!<%?$r%"%C%W%m!<%I$7$F$$$^$9!#(B
> 
> For that reason, in the case of communicating inadvisable commercial use, in the comments explicit contact has to be made for data uploaded to Thingiverse.

(Note: this is really tough.  I'm only 50% sure of the above translation.  There are so many interpretations here... I think I had better ask a Japanese colleague... will get back to you)

Michael

> On 2015/03/15, at 0:41, Michael McCool michael.d.mccool@gmail.com wrote:
> 
> $BF|K\?M$H$7$F$O!"CQ$:$+$7$$$N$G$9$,!"8=:_$NF|K\@=$N(B3D$B%W%j%s%?$K1w$$$F!"(Bmarlin$B$,;HMQ$5$l$F$$$J$,$i$=$l$,L@<($5$l$F$$$J$$J_$d%=!<%9%3!<%I$r8x3+$7$F$$$J$$J_$,B8:_$7$F$$$^$9!#(B
> 
> As for Japanese people, it's embarassing but, in modern Japanese-made printers, while Marlin is being used that fact is not made explicit, and the existence of the source code is not made public.
> 
> > On 2015/03/15, at 0:30, Michael McCool michael.d.mccool@gmail.com wrote:
> > 
> > Let me give it a stab (I live in Japan; my Japanese is not super-fluent, though).
> > 
> > > $B1Q8l$,6l<j$J$N$GK]Lu%5%$%H$rMxMQ$7$F$$$^$9!#(BBecause my English is bad I am using a translated web site.  
> > > 
> > > $BF|K\8l2=$7$?(Bmarlin$B$K$D$$$F$G$9!#(BRegarding the Japanese language translation of Marlin...
> > > 
> > > More to follow, next sentence is a tough run-on one, need to dig out the dictionary, too.  It's past midnight here though so I'll tackle it in the morning.  Something about wanting to use Marlin in Japanese-made 3D printers....
> > 
> > Michael McCool
> > 
> > > On 2015/03/14, at 21:27, Scott Lahteine notifications@github.com wrote:
> > > 
> > > @AnHardt Well$B!D(B Here's what I have. I need help in translation and communication$B!D(B
> > > 
> > > $B1Q8l$,6l<j$J$N$GK]Lu%5%$%H$rMxMQ$7$F$$$^$9!#(B
> > > 
> > > $BF|K\8l2=$7$?(Bmarlin$B$K$D$$$F$G$9!#(B
> > > $BF|K\?M$H$7$F$O!"CQ$:$+$7$$$N$G$9$,!"8=:_$NF|K\@=$N(B3D$B%W%j%s%?$K1w$$$F!"(Bmarlin$B$,;HMQ$5$l$F$$$J$,$i$=$l$,L@<($5$l$F$$$J$$J_$d%=!<%9%3!<%I$r8x3+$7$F$$$J$$J_$,B8:_$7$F$$$^$9!#(B
> > > 
> > > $B$=$N0Y!">&MQMxMQIT2D$N5-:\$H$7$F!"%3%a%s%HJ8$K>&MQMxMQ$9$k>l9g$OO"Mm$9$kMM$KL@<($7$F(BThingiverse$B$X%G!<%?$r%"%C%W%m!<%I$7$F$$$^$9!#(B
> > > 
> > > $B>&MQMxMQ$NLd$$9g$;$,M-$C$?>l9g$K$O!"(Bmarlin$B$,;HMQ$5$l$F$$$k;v$NL@<($H%=!<%9%3!<%I$N3+<($r9T$($P5v2D$7$^$9$N$G!"<B<AE*$K$O>&MQMxMQ2DG=$H$9$k$D$b$j$G$9!#(B
> > > 
> > > $B!=(B
> > > Reply to this email directly or view it on GitHub.

AnHardt commented:
@mmccool 
A big THANKS for now.

mmccool commented:
Here's the translation provided by my colleagues.  However if there is going to be a lot of to-and-fro with this person I may want to ask my colleagues to get directly involved.  We work in IoT in Japan with Intel so supporting the development of open-source 3D printer software is not completely out of the question.   I actually have a bunch in the lab I am working on ;)

Michael McCool

Pardon me for using a machine translating site. Ifm not good at communicating in English.

This email is about gMarlinh used in Japanese language.
There exist some 3D printers made in Japan which do not exhibit use of Marlin or do not open the source code, (which is a shame as one of Japanese).

I have explicitly commented on Thingiverse that Marlin cannot be used for commercial purpose AND those who wish to use Marlin for the purpose should contact us.

We plan to approve commercial use, as long as user agrees to 1) explicitly notify the use of Marlin, and 2) open the source code. This means, practically, we intend to make it possible to use Marlin for commercial purpose.

> On 2015/03/15, at 1:14, AnHardt notifications@github.com wrote:
> 
> @mmccool 
> A big THANKS for now.
> 
> \
> Reply to this email directly or view it on GitHub.

thinkyhead commented:
:+1: Much better results than before, thank you! Well that is pretty much consistent with GPLv3. Vendors are supposed to provide the source if they use Marlin, and he is just asserting GPLv3 in a certain way. But this might fall just a little short of receiving permission to adapt his work for the main branch of Marlin, so maybe it would help if he could talk to a native-speaker about gaining that permission. He is user "ichibey" on Thingiverse.

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

AnHardt commented:
@mmccool 
Thank you and your colleagues.
That's not what we wanted to hear but what we expected.

@thinkyhead 
I'll prepare a version without the 'kana' font. Throwing it in later is no problem when the supporting structures are in the code.

@mmccool & colleagues
Would a pure katakana version be of any help for the average Japanese 3D-printer user?
If the answer is yes - would you pleas be so kind to check: https://github.com/AnHardt/Marlin/blob/utf/Marlin/language_kana_utf8.h
for me.
This is the UTF-version i derived from ichibey's https://github.com/AnHardt/Marlin/blob/utf/Marlin/language_kanji.h.
Replaced the \xxx sequences by (for me) similar looking UTF8 symbols but have abslolue no idea if this was correct. How you translation would look like?

AnHardt commented:
Sorry. Edited last message.
Was derived from: https://github.com/AnHardt/Marlin/blob/utf/Marlin/language_kana.h

boelle commented:
Hmm... have we asked him very direct?

ie: Can we use your work in the marlin firmware?

thinkyhead commented:
@boelle Well in my communication I hoped that I could make it clear that if he says "ok" then we have to remove the commercial use restriction. That said, he still has the right under GPLv3 to request a fee for commercial use.

boelle commented:
and if we include his work we would have the trouble of handling that right? or should the facotry that decides to use marlin have to deal with that?

thinkyhead commented:
@boelle Well the GPLv3 is about user rights. If a user wants the source code from a vendor, they can assert this right and win in court. So really it's no extra use to add a commercial restriction on the basis that vendors aren't releasing the code. I can't even say that Ichibey has put that to the test. It may just be his opinion that vendors are not releasing the code, when really he just never bothered to ask loudly enough.

boelle commented:
So...

1) we should ask him again and ask for a NO or YES...

2) if he says yes but still want a fee if the firmware is used commercial... who has the trouble of collecting the fee?

thinkyhead commented:
He would be responsible for collecting his fee. Marlin documentation can mention this fee and link to his site. I don't know that he needs a fee, that is just an example of a right that he has which doesn't violate GPLv3.

boelle commented:
ok...

So to make it short we just need a clear yes or no from him?

and if he want a link to his site etc?

AnHardt commented:
Ichibey never asked for a fee, just for his allowance.

boelle commented:
oki.... so we just need a clear yes or no then....

mmccool commented:
I would just ask him some very simple, direct questions in clear English.  Avoid idioms.   In that case machine translation can do a good job.   Then ask him to respond using very simple sentences.   If he avoids long run-on sentences and complex conjuctions, and is careful to be explicit about things, then again machine translation can work... although, in my experience, Japanese to English is much harder than the other way around.   It's not so much the language but the way it is used...

If you have trouble interpreting his response ping me again.

Michael

> On 2015/03/15, at 23:54, Bo Herrmannsen notifications@github.com wrote:
> 
> oki.... so we just need a clear yes or no then....
> 
> \
> Reply to this email directly or view it on GitHub.

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream changes

thinkyhead committed:
Get upstream commits

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream commits.

thinkyhead commented:
@AnHardt By "allowance" do you mean "permission" or… "name in the credits" or… something else? Anyway, I'll need to be redo this a bit, since the language pack is now merged. Then I will see about clarifying things with Ichibey, for which I might want to recruit a native-speaker.

thinkyhead committed:
Broken hybrid – with Big Language Pack

- My original code in `dogm_lcd_implementation.h` commented
- The original language_jp files included

thinkyhead commented:
@AnHardt This PR is now a hybrid, with my old version of the Kana integration commented out in the code, and with the old-named files that I originally added hanging around, some or all of which are superseded by your updated files. The commented-out parts I will replace with the appropriate code, if needed, with changes to include paths, and the old font and language files I can simply remove if they are totally now obsolete.

AnHardt commented:
@thinkyhead 
What you call Kana here is what the original author called his Kanji version. Anyway you know what we are talking about.
I have built in a recipe to integrate Kanji again. When needed just revert https://github.com/MarlinFirmware/Marlin/commit/2f374f146537387e415b525782e27d50beae42c8.
I hope this will work out of the box then.

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream commits

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream commits

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream commits

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream commits

thinkyhead committed:
Merge branch 'Development' into japanese_kana_language
Latest upstream commits

boelle commented:
@thinkyhead you marked it as need update, but it can be merged automatic?

thinkyhead commented:
@boelle It's compatible with the source code, but not in a state which should be merged. The right way to get Japanese support enabled is to revert https://github.com/MarlinFirmware/Marlin/commit/2f374f146537387e415b525782e27d50beae42c8 as @AnHardt pointed out. I will close this, as it is meant only to ask the question, "what needs to be done to get Japanese working?"

AnHardt commented:
@esenapaj 
If you are bored, you could have a look on this.
If a Japanese Kanji version could have any advantages over the Kana version, you are the one who has all the needed abilities now, to realize it. 
A working example for this kind of mapping is the cn-version.

Just an idea.

ghost commented:
@AnHardt 
Thank you for letting me know.
I had been aware of that resources of kanji removed in the past,
but I hadn't been aware of that it was removed for what reason.

And I began (re-)implementation kanji(kanji and kana), but it's bit troublesome.
I estimated the number of letters that  those are at least the required, and it's over 128 letters.
https://github.com/esenapaj/Marlin/blob/kanji_kana/Marlin/language_kanji_and_kana.h

```
hiragana: 23 letters

あいおかがきくさしせたちてにはへまみめら
りれん
```

```
katakana: 39 letters

ァイィエオォカクグコサシスズセタダッデト
ドノバピフブプベミメモュラリルレロンー
```

```
kanji: 85 letters

移一引押温下加過開格較完換基帰機径原源後
高交込再最作止始自時待軸失取出準心小冗常
正整設全暫走送挿速退大中長調超直定低停適
点電度動読入熱納能敗非避備微復分暴間満未
躍予用了量
```

```
total: 147 letters
```

If I attempt to realize to re-implement kanji, I have to curtail the needed letters in one way or another...
